### DISCUSSION TITLE
Allow direct casting between type parameters and all types

### DISCUSSION THREAD

**Discussion by ekolis** [dotnet/csharplang#8812] [Language Ideas]
Let's say I have this:

```
public string DoStuff<T>(T t)
{
	if (t is decimal)
	{
		decimal d = (decimal)t;
		Console.WriteLine("the number is: " + d);
	}
}
```

This code does not compile because `T` cannot be cast directly to `decimal`; I have to do an intermediate cast to `object` (`void*` much?!):

```
decimal d = (decimal)(object)t;
```

Seems rather unnecessary to me... why can't I cast the type parameter instance directly to whatever type I want?

**Comment by HaloFour** [comment]
The compiler can't validate the cast.  The fact that you did an `is` check on the previous line does not impact the analysis of the compiler.

In this case you can use pattern matching and eliminate the need for the cast:

```cs
public string DoStuff<T>(T t)
{
	if (t is decimal d)
	{
		Console.WriteLine("the number is: " + d);
	}
}
```

**Comment by IS4Code** [comment]
Interestingly enough, `is decimal d` *boxes* the value, so it is in essence equivalent to obtaining `(object)t as decimal`. I can see now why the runtime specifically tries to optimize this situation and get rid of actual boxing.
